File: src/zoneA/tuning.ts
/**
 * Zone A gameplay tuning — plain numbers only, no Phaser, no logic.
 *
 * Lives in `zoneA/` (not the shared `core/Layout.ts`) because this is Dev 1's
 * Zone-A feel, tweakable without touching the seam. `ballMath.ts` turns these into
 * the per-tier values the system reads at runtime.
 *
 * Design space is Layout's 390x1238 world; Zone A is the top 390x507 band (42px HUD +
 * 465px board — ~60% of the 844px screen). In the B phase the camera top-crops the board
 * to 71px on screen (HUD + board = a 113px sliver), but these numbers are world-space
 * and never change with the phase framing.
 */

// --- Geometry (design-space px) -------------------------------------------

/** The row the current/aim ball sits on, near the top. Above the death line and far
 *  enough below the HUD chrome bar (42px tall) that the largest spawnable ball (tier 4,
 *  radius 34) clears it: top edge at y=44 clears the bar, bottom at y=112 stays above the
 *  death line. */
export const SPAWN_Y = 78;

/** Forgiving death line, just below the spawn row: a ball resting ABOVE this for
 *  REST_MS ends the run (see ballMath.isRestingAbove / isOverflow). In the B-phase
 *  framing this row is cropped off-screen — benign, since only sucks (removals) touch
 *  the board then, so it can't newly overflow while invisible. */
export const DEATH_LINE_Y = 108;

/** Ball radius for the base tier table; index = tier-1. Has TIER_COUNT (10) entries.
 *  tier-10 diameter 198 < 390 (fits the base arena); tier-4 radius 34 < SPAWN_Y (clears ceiling).
 *  Low tiers are deliberately chunky (~30% over the original table) so a 12-ball buffer
 *  visibly crowds the base board — early-game tension comes from board pressure, not count.
 *  Merges are uncapped, so tiers past the table grow geometrically — see RADIUS_GROWTH. */
export const RADII: readonly number[] = [17, 22, 28, 34, 41, 50, 60, 71, 84, 99];

/** Per-tier radius multiplier beyond the base table (≈ the table's own top step, 78/66).
 *  radiusForTier(tier > 10) = RADII[last] * RADIUS_GROWTH^(tier-10). */
export const RADIUS_GROWTH = 1.18;

// --- Physics --------------------------------------------------------------

/** Surface friction grows with tier: FRICTION_BASE + FRICTION_STEP*(tier-1), clamped. */
export const FRICTION_BASE = 0.4;
export const FRICTION_STEP = 0.025;
export const FRICTION_MAX = 0.5;

/** Constant friction terms applied to every ball. */
export const FRICTION_AIR = 0.01;
export const FRICTION_STATIC = 0.1;

/** Surface friction of the funnel floor. Matter combines contacts as
 *  min(ballFriction, floorFriction), so this floor value governs how readily
 *  balls slide — lower = smaller/lighter balls slide toward the apex more. */
export const FLOOR_FRICTION = 0.02;

/** Base density for the small tiers — Matter derives mass from density*area. Larger balls
 *  taper below this (see DENSITY_TAPER_TIER / DENSITY_MASS_EXP) so their mass — and the
 *  collision momentum between big balls — grows sub-quadratically instead of ∝ r². */
export const DENSITY = 0.02;

/** Tiers at or below this keep the flat DENSITY (early/mid-game balance untouched); larger
 *  tiers taper. `densityForTier` in ballMath.ts turns this into the per-tier density. */
export const DENSITY_TAPER_TIER = 8;

/** Above the taper tier, mass grows like radius^DENSITY_MASS_EXP (was ∝ r², i.e. exp 2). 1 =
 *  linear in radius, so a tier-20 ball ends up ~5× lighter than a flat-density one would be. */
export const DENSITY_MASS_EXP = 1;

/** Restitution (bounciness). Modest, so balls settle but still bounce/roll a little. */
export const RESTITUTION = 0.2;

// --- Collision categories -------------------------------------------------

/**
 * Matter collision-filter bitmask categories for Zone A's own bodies. Zone B owns
 * `0x0001`–`0x0008` (see `zoneB/ZoneBBall.ts`), and Zone B balls mask exactly those bits —
 * so Zone A takes the *next* bits, which Zone B balls never collide with. This is what keeps
 * Zone A's boundary walls + funnel (whose thickness now scales with the arena, so at the last
 * milestone the funnel reaches hundreds of px past the Zone A/C seam into Zone B's space) from
 * catching balls that belong to Zone B. Zone A balls and walls collide only with each other,
 * never with anything in Zone B. Keep these outside the `0x0001`–`0x0008` range Zone B uses.
 */
export const CAT_ZONE_A_BALL = 0x0010;
export const CAT_ZONE_A_WALL = 0x0020;

// --- Merge blast ----------------------------------------------------------

/** Neighbours within this radius of a merge get nudged outward. Base value at arena
 *  scale 1 — Board multiplies it by the live scale so the reach tracks ball sizes. */
export const BLAST_RADIUS = 60;

/** Peak outward velocity kick at the merge point (linear falloff to 0 at the radius).
 *  Base value at arena scale 1, scaled like BLAST_RADIUS. */
export const BLAST_STRENGTH = 1.6;

// --- Overflow / game over -------------------------------------------------

/** A body whose `speed` is below this counts as "at rest". Base value at arena scale 1 —
 *  Board scales it by the live scale (normalized gravity makes world speeds grow with it). */
export const REST_SPEED = 0.8;

/** Hard ceiling on a ball's per-step speed (px/step at arena scale 1); Board scales it by the
 *  live scale, so it tracks the ×s world speeds. The anti-tunnel backstop: it MUST stay below
 *  WALL_T (the ArenaView wall thickness, also ×s), so a ball can never cross a wall in one step
 *  even from stacked merge-blasts. Set well above normal play speeds (rest 0.8, blast peak 1.6)
 *  so it only clips runaway cases and doesn't change ordinary feel. */
export const MAX_BALL_SPEED = 16;

/** How long a ball must rest above the death line before the run ends (ms). */
export const REST_MS = 1000;

/** Px below the death line within which a resting ball flags the red warning line. */
export const WARN_BAND = 28;

// --- Input ----------------------------------------------------------------

/** Brief lock-out after a drop so the next ball can't be slammed into the same spot. */
export const DROP_COOLDOWN_MS = 250;

// --- Colour ---------------------------------------------------------------

// The tier look now lives in the shared `core/Materials` ladder (name + colours +
// physics feel per tier), so Zone A and Zone B stay pixel-identical — a transferred
// ball keeps its material. Import `materialForTier`/`colorForTier` from there.

File: Assets/Game/Core/Tuning.cs
namespace RichCoast.Core
{
    /// <summary>
    /// Zone A gameplay tuning — plain numbers only, no logic. Ported verbatim from the original
    /// <c>zoneA/tuning.ts</c>; <see cref="BallMath"/> turns these into the per-tier values the
    /// system reads at runtime.
    ///
    /// Design space is Layout's 390×1238 world; Zone A is the top 390×507 band (42 HUD +
    /// 465 board). In the B phase the camera top-crops the board to a 71 sliver, but these
    /// numbers are world-space and never change with the phase framing.
    /// </summary>
    public static class Tuning
    {
        // --- Geometry (design-space units) -----------------------------------

        /// <summary>
        /// The row the current/aim ball sits on, near the top. Above the death line and far
        /// enough below the 42-tall HUD bar that the largest spawnable ball (tier 4, radius 34)
        /// clears it.
        /// </summary>
        public const float SpawnY = 78f;

        /// <summary>
        /// Forgiving death line, just below the spawn row: a ball resting ABOVE this for
        /// <see cref="RestMs"/> ends the run. In the B-phase framing this row is cropped
        /// off-screen — benign, since only removals touch the board then.
        /// </summary>
        public const float DeathLineY = 108f;

        /// <summary>
        /// Base tier radii; index = tier-1, <see cref="Tiers.TierCount"/> entries. Tier-10
        /// diameter 198 &lt; 390 (fits the base arena); tier-4 radius 34 &lt; SpawnY (clears the
        /// ceiling). Low tiers are deliberately chunky so a small buffer visibly crowds the base
        /// board — early-game tension comes from board pressure, not ball count.
        /// </summary>
        public static readonly float[] Radii = { 17f, 22f, 28f, 34f, 41f, 50f, 60f, 71f, 84f, 99f };

        /// <summary>Per-tier radius multiplier beyond the base table (≈ the table's own top step).</summary>
        public const float RadiusGrowth = 1.18f;

        // --- Physics ---------------------------------------------------------

        /// <summary>Surface friction grows with tier: Base + Step*(tier-1), clamped to Max.</summary>
        public const float FrictionBase = 0.4f;
        public const float FrictionStep = 0.025f;
        public const float FrictionMax = 0.5f;

        /// <summary>Constant friction terms applied to every ball.</summary>
        public const float FrictionAir = 0.01f;
        public const float FrictionStatic = 0.1f;

        /// <summary>
        /// Surface friction of the funnel floor. Contacts combine as min(ball, floor), so this
        /// governs how readily balls slide — lower = smaller balls slide toward the apex more.
        /// </summary>
        public const float FloorFriction = 0.02f;

        /// <summary>
        /// Base density for the small tiers — mass derives from density×area. Larger balls taper
        /// below this so collision momentum between big balls grows sub-quadratically.
        /// </summary>
        public const float Density = 0.02f;

        /// <summary>Tiers at or below this keep the flat <see cref="Density"/>; larger tiers taper.</summary>
        public const int DensityTaperTier = 8;

        /// <summary>
        /// Above the taper tier, mass grows like radius^this (was ∝ r², i.e. 2). 1 = linear in
        /// radius, so a tier-20 ball ends up ~5× lighter than a flat-density one would be.
        /// </summary>
        public const float DensityMassExp = 1f;

        /// <summary>Restitution (bounciness). Modest, so balls settle but still bounce/roll a little.</summary>
        public const float Restitution = 0.2f;

        /// <summary>
        /// Physics steps per second. The original tuning is authored per STEP (a 60 Hz Matter.js
        /// step), so the per-step speed constants below convert to per-second by this factor.
        /// The fixed timestep is pinned to match.
        /// </summary>
        public const float StepsPerSecond = 60f;

        /// <summary>
        /// Downward acceleration in design units/s². The original relied on Matter.js's implicit
        /// gravity scaling, which has no Unity equivalent, so this is set by feel against the
        /// authored geometry: a ball released at the spawn row reaches the funnel floor of the base
        /// arena in a little over half a second. It is the one Zone A number that is genuinely new
        /// rather than ported, and the first knob to turn if the drop feels off.
        /// </summary>
        public const float Gravity = 2000f;

        /// <summary>Rest threshold in design units/second — <see cref="RestSpeed"/> per step.</summary>
        public const float RestSpeedPerSecond = RestSpeed * StepsPerSecond;

        /// <summary>Anti-tunnel speed ceiling in design units/second — <see cref="MaxBallSpeed"/> per step.</summary>
        public const float MaxBallSpeedPerSecond = MaxBallSpeed * StepsPerSecond;

        // --- Merge blast -----------------------------------------------------

        /// <summary>
        /// Neighbours within this radius of a merge get nudged outward. Base value at arena scale
        /// 1 — the board multiplies it by the live scale so the reach tracks ball sizes.
        /// </summary>
        public const float BlastRadius = 60f;

        /// <summary>Peak outward kick at the merge point (linear falloff to 0 at the radius).</summary>
        public const float BlastStrength = 1.6f;

        // --- Overflow / game over --------------------------------------------

        /// <summary>
        /// A body slower than this counts as "at rest". Base value at arena scale 1 — the board
        /// scales it by the live scale, since normalized gravity makes world speeds grow with it.
        /// </summary>
        public const float RestSpeed = 0.8f;

        /// <summary>
        /// Hard ceiling on a ball's per-step speed at arena scale 1, scaled by the live scale.
        /// The anti-tunnel backstop: it MUST stay below the arena wall thickness (also scaled), so
        /// a ball can never cross a wall in one step even from stacked merge-blasts. Set well
        /// above normal play speeds so it only clips runaway cases.
        /// </summary>
        public const float MaxBallSpeed = 16f;

        /// <summary>How long a ball must rest above the death line before the run ends (ms).</summary>
        public const float RestMs = 1000f;

        /// <summary>Distance below the death line within which a resting ball flags the red warning line.</summary>
        public const float WarnBand = 28f;

        // --- Zone B ----------------------------------------------------------

        /// <summary>
        /// Zone B ball radius. Deliberately tiny: many balls coexist in the cascade, and at this
        /// size colour alone carries a ball's identity.
        /// </summary>
        public const float ZoneBBallRadius = 10f;

        /// <summary>Physical thickness of a gate bar.</summary>
        public const float GateThickness = 16f;

        /// <summary>Zone B surface feel, before each material's own multipliers.</summary>
        public const float ZoneBRestitution = 0.35f;

        /// <summary>Cap on restitution after the material multiplier, so exotic tiers stay lively
        /// without ping-ponging the cascade forever.</summary>
        public const float ZoneBRestitutionMax = 0.5f;

        public const float ZoneBFriction = 0.05f;
        public const float ZoneBFrictionAir = 0.008f;

        /// <summary>
        /// Freshly-split balls ignore gates briefly, or a copy born touching the bar it was just
        /// split by would split again immediately and the cascade would run away.
        /// </summary>
        public const float SplitGraceMs = 300f;

        /// <summary>Angular spread of a split fan (radians) and how far copies are offset from the hit.</summary>
        public const float SplitSpread = 0.8f;
        public const float SplitOffset = 18f;

        /// <summary>Launch speed of a split copy, per step at 60 Hz.</summary>
        public const float SplitSpeed = 3f;

        /// <summary>
        /// Hard cap on balls in flight. A deep cascade is the point, but a budget phone cannot
        /// simulate an unbounded one — beyond this a split simply stops adding copies.
        /// </summary>
        public const int ZoneBMaxBalls = 200;

        /// <summary>Levels a single cash-in may roll through before the excess is forfeited.</summary>
        public const int MaxLevelsPerCashIn = 10;

        // --- Input -----------------------------------------------------------

        /// <summary>Brief lock-out after a drop so the next ball can't be slammed into the same spot.</summary>
        public const float DropCooldownMs = 250f;
    }
}
